### NatVar_proj
This holds scripts and data associated with the manuscript describing natural flg22 variation.

# ALL SCRIPTS ARE AS IS
You will likely have to change file pointers and directory call outs for use on your local computer. Please email ncolaian@live.unc.edu for any questions or inquiries on the data or code contained here.

The database containing protein, dna, gff, and other for Arabidopsis associated bacterial strains can be found at:

The database containing the NCBI protein filies used to identify Ralstonia FliC and flg22 varriants can be found at:

The ROS folder contains all the data and code used to make all ROS burst figures in the publication.

The SGI folder contains all the sgi data and analysis code. It also contains the combined ROS and SGI figures for the Ralstonia peptides and the rest of the peptides that were screened.

The FliC_pull_code contains scripts, files, and representative commands used to identify fliC and flg22 variants. These scripts and examples could be used to identify flg22 variants in any genome/database of interest

The supp_dir contains all the supplemental files updated to the manuscript.

The fliC_flg22_functionality directory contains the raw data and scripts used to functionally annotate the flg22 variants found in the synthetic communities reported by Finkel et al., 2020 ( https://doi.org/10.1038/s41586-020-2778-7 )

The fliC_clust_arabonly directory  contains information concerning fliC cluster identification

The flg22_phylo directory contains metadata and newick file used to create figure 2C

The confocal_data directory contains the confocal data and script used to create the figures associated with connfocal analysis (Figures - )

The cl21_exps directory contains the data and code used to analyze the effect of fliC replacement on motility and root growth inhibition.

The ELISA directory contains the data and statistics performed on the ELISA absorbance readings displayed in the manuscript


